### ISSUE TITLE
No Index Checking in Dataset::CopySubset 

### ISSUE THREAD

**Issue by Tony-Y** [Microsoft/LightGBM#959]
I noticed no index checking exists in Dataset::CopySubset. I think this function must have a code such as the following:

```c++
void Dataset::CopySubset(const Dataset* fullset, const data_size_t* used_indices, data_size_t num_used_indices, bool need_meta_data) {
    for (data_size_t i = 0; i < num_used_indices; ++i) { 
        if (used_indices[i] < 0 || used_indices[i] >= fullset->num_data_) {
            Log::Fatal("At least, one of used indices is out-of-bounds.");
            is_finish_load_ = false;
            return
        }
    } 
}
```

**Comment by Tony-Y** [comment]
I show a simple python example throwing no error.

# test_subset.py
```python
from __future__ import print_function
import lightgbm as lgb
import pandas as pd

df_train = pd.read_csv('../regression/regression.train', header=None, sep='\t')
y_train = df_train[0].values
X_train = df_train.drop(0, axis=1).values

lgb_train = lgb.Dataset(X_train, y_train).construct()
n = lgb_train.num_data()
print(n)

used_indices = range(1,n+1)
lgb_subset = lgb_train.subset(used_indices).construct()
m = lgb_subset.num_data()
print(m)
```

# How to run this code
cd LightGBM/examples/python-guide
python test_subset.py

# output
7000
7000

**Comment by guolinke** [comment]
@Tony-Y you can submit a PR to fix this. 
However, I prefer to check it in C_API (https://github.com/Microsoft/LightGBM/blob/master/src/c_api.cpp#L674-L694), since we will use this function inner and checking this will cause the performance issue.

**Comment by Tony-Y** [comment]
@guolinke I submitted a PR.
I confirmed the bounds checking works well using the following python script: 

### test_subset.py
```python
from __future__ import print_function
import lightgbm as lgb
import pandas as pd
import numpy as np

df_train = pd.read_csv('../regression/regression.train', header=None, sep='\t')
y_train = df_train[0].values
X_train = df_train.drop(0, axis=1).values

lgb_train = lgb.Dataset(X_train, y_train).construct()
n = lgb_train.num_data()
print(n)

np.random.seed(3691)

def shift(k,use_shuffle=False):
  print("<Shift: {0}>".format(k))
  try:
    used_indices = np.arange(k,n+k)
    if use_shuffle:
      np.random.shuffle(used_indices)
    lgb_subset = lgb_train.subset(used_indices).construct()
    m = lgb_subset.num_data()
    print(m)
  except lgb.basic.LightGBMError as err:
    print("Catched a LightGBMError")
    print("Message: {0}".format(err))

print("No shuffle")
shift(0)
shift(-1)
shift(-2)
shift(1)
shift(2)

print("Use shuffle")
shift(0,True)
shift(-1,True)
shift(-2,True)
shift(1,True)
shift(2,True)
```

### Output
```
7000
No shuffle
<Shift: 0>
7000
<Shift: -1>
[LightGBM] [Fatal] [Used indices of subset]: does not tolerate element [#0 = -1] outside [0, 6999]
Catched a LightGBMError
Message: b'[Used indices of subset]: does not tolerate element [#0 = -1] outside [0, 6999]'
<Shift: -2>
[LightGBM] [Fatal] [Used indices of subset]: does not tolerate element [#0 = -2] outside [0, 6999]
Catched a LightGBMError
Message: b'[Used indices of subset]: does not tolerate element [#0 = -2] outside [0, 6999]'
<Shift: 1>
[LightGBM] [Fatal] [Used indices of subset]: does not tolerate element [#6999 = 7000] outside [0, 6999]
Catched a LightGBMError
Message: b'[Used indices of subset]: does not tolerate element [#6999 = 7000] outside [0, 6999]'
<Shift: 2>
[LightGBM] [Fatal] [Used indices of subset]: does not tolerate element [#6998 = 7000] outside [0, 6999]
Catched a LightGBMError
Message: b'[Used indices of subset]: does not tolerate element [#6998 = 7000] outside [0, 6999]'
Use shuffle
<Shift: 0>
7000
<Shift: -1>
[LightGBM] [Fatal] [Used indices of subset]: does not tolerate element [#2425 = -1] outside [0, 6999]
Catched a LightGBMError
Message: b'[Used indices of subset]: does not tolerate element [#2425 = -1] outside [0, 6999]'
<Shift: -2>
[LightGBM] [Fatal] [Used indices of subset]: does not tolerate element [#2970 = -2] outside [0, 6999]
Catched a LightGBMError
Message: b'[Used indices of subset]: does not tolerate element [#2970 = -2] outside [0, 6999]'
<Shift: 1>
[LightGBM] [Fatal] [Used indices of subset]: does not tolerate element [#944 = 7000] outside [0, 6999]
Catched a LightGBMError
Message: b'[Used indices of subset]: does not tolerate element [#944 = 7000] outside [0, 6999]'
<Shift: 2>
[LightGBM] [Fatal] [Used indices of subset]: does not tolerate element [#41 = 7000] outside [0, 6999]
Catched a LightGBMError
Message: b'[Used indices of subset]: does not tolerate element [#41 = 7000] outside [0, 6999]'
```

**Comment by StrikerRUS** [comment]
Fixed in #969.
